File: validsudoku.js
//Objective is to see if a sudoku puzzle is valid


//O(1) solution since we are always checking a 9x9 board

let rows = new Set()
let cols = new Set()
let boxes = new Set()
    
for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
        let currRowElement = board[i][j]
        let currColElement = board[j][i]
        let currBoxElement = board[3 * Math.floor(i / 3) + Math.floor(j / 3)][((i * 3) % 9) + (j % 3)]
        
        //Make sure the current row, column, and box all don't have the current element
        if (rows.has(currRowElement)) {
            return false
        }
        if (currRowElement !== '.') {
            rows.add(currRowElement)
        }
            
        if (cols.has(currColElement)) {
            return false
        }
        if (currColElement !== '.') {
            cols.add(currColElement)
        }
            
        if (boxes.has(currBoxElement)) {
            return false
        }
        if (currBoxElement !== '.') {
            boxes.add(currBoxElement)
        }
    }
    //Clear everything once we enter a new box
    rows.clear()
    cols.clear()
    boxes.clear()
}
    
    return true

File: validsudoku.java
//Java Solution

class Solution {
    public boolean isValidSudoku(char[][] board) {
        Set<Character> rows = new HashSet<>();
        Set<Character> cols = new HashSet<>();
        Set<Character> boxes = new HashSet<>();
        
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                char rowElement = board[i][j];
                char colElement = board[j][i];
                char boxElement = board[(int) (3 * Math.floor(i / 3)) + (int) Math.floor(j / 3)][((i * 3) % 9) + (j % 3)];
                
                if (rows.contains(rowElement)) {
                    return false;
                }
                if (rowElement != '.') {
                    rows.add(rowElement);
                }
                
                if (cols.contains(colElement)) {
                    return false;
                }
                if (colElement != '.') {
                    cols.add(colElement);
                }
                
                if (boxes.contains(boxElement)) {
                    return false;
                }
                if (boxElement != '.') {
                    boxes.add(boxElement);
                }
            }
            
            rows.clear();
            cols.clear();
            boxes.clear();
        }
        
        return true;
    }
}